Python/Product/PythonTools/PythonTools/LanguageServerClient/FileWatcher/Listener.cs

```diff
@@ -108,72 +108,91 @@ class Listener : IDisposable {
-            if (!e.IsDirectoryChanged() && _rpc != null) {
-                // Create something to match with
-                var item = new InMemoryDirectoryInfo(_root, new string[] { e.FullPath });
-
-                // See if this matches one of our patterns.
-                if (_matcher.Execute(item).HasMatches) {
-                    // Send out the event to the language server
-                    var renamedArgs = e as System.IO.RenamedEventArgs;
-                    var didChangeParams = new DidChangeWatchedFilesParams();
-
-                    // Visual Studio actually does a rename when saving. The rename is from a file ending with '~'
-                    if (renamedArgs == null || renamedArgs.OldFullPath.EndsWith("~")) {
-                        renamedArgs = null;
-                        didChangeParams.Changes = new FileEvent[] { new FileEvent() };
-                        didChangeParams.Changes[0].Uri = new Uri(e.FullPath);
-
-                        switch (e.ChangeType) {
-                            case WatcherChangeTypes.Created:
-                                didChangeParams.Changes[0].FileChangeType = FileChangeType.Created;
-                                break;
-                            case WatcherChangeTypes.Deleted:
-                                didChangeParams.Changes[0].FileChangeType = FileChangeType.Deleted;
-                                break;
-                            case WatcherChangeTypes.Changed:
-                                didChangeParams.Changes[0].FileChangeType = FileChangeType.Changed;
-                                break;
-                            case WatcherChangeTypes.Renamed:
-                                didChangeParams.Changes[0].FileChangeType = FileChangeType.Changed;
-                                break;
-
-                            default:
-                                didChangeParams.Changes = Array.Empty<FileEvent>();
-                                break;
-                        }
-                    } else {
-                        // file renamed
-                        var deleteEvent = new FileEvent();
-                        deleteEvent.FileChangeType = FileChangeType.Deleted;
-                        deleteEvent.Uri = new Uri(renamedArgs.OldFullPath);
 
-                        var createEvent = new FileEvent();
-                        createEvent.FileChangeType = FileChangeType.Created;
-                        createEvent.Uri = new Uri(renamedArgs.FullPath);
+            // Skip directory change events
+            if (e.IsDirectoryChanged()) {
+                return;
+            }
 
-                        didChangeParams.Changes = new FileEvent[] { deleteEvent, createEvent };
+            // If the rpc has been disconnected, the _rpc_Disconnected callback will set _rpc to null.
+            // So check for null now, and also check before every rpc call to avoid a null reference exception
+            if (_rpc is null) {
```

Comment:
I think it would be better to copy it into a local - that way there's no assignment race, and you only need to worry about caching the exception from `Dispose`.

Reply:
So you're saying avoid the null check altogether and just see what exception gets thrown when trying to use a disposed rpc, and catch that along with the one I already caught?

Reply:
It can be null to begin with, so I don't see any way to avoid the check. But if you assign to a local instead, it could be just a single check on that local.